### PR TITLE
Migrate docs to Antora site

### LINKED ISSUE #975: Migrate docs to Antora site
As mentioned in https://github.com/asciidoctor/asciidoctor.org/issues/977, actions have begun for the new site current pre-release in https://asciidoctor-docs.netlify.app/asciidoctor/2.0/.

I finished migration of the maven-plugin docs (open for review https://github.com/asciidoctor/asciidoctor-maven-plugin/pull/498), so if there're no objection I will start work here too.

Given the Asciidoctor site is scheduled to release soon I'll only focus on docs, no gh-pages.

### PR DESCRIPTION
**Pull request by abelsromero**
Thank you for opening a pull request and contributing to AsciidoctorJ!

Please take a bit of time giving some details about your pull request:

## Kind of change

- [ ] Bug fix
- [ ] New non-breaking feature
- [ ] New breaking feature
- [x] Documentation update
- [ ] Build improvement

## Description

What is the goal of this pull request?

Migrate docs to Antora site as discussed in #975 

How does it achieve that?

Splits README and current documents under _docs_ into modules and pages following Antora patterns.

Are there any alternative ways to implement this?
N/A

Are there any implications of this pull request? Anything a user must know?

Regardig the PR, **untill said otherwise, it's WIP**.
Given this is a buig change I think it's better to to push changes frequently so the changes can be reviewed one by one.
At the end I will squash.

NOTE: On top of breaking the files are minnor changes "here and there" hard to spot that make sense once files are split.
For example, the introduction below the `Developmentp` title has been summarized and added to the intro of `Project layout`.
I'll try to keep to minimal though.

## Issue

If this PR fixes an open issue, please add a line of the form:

Fixes #975 

## Release notes

Please add a corresponding entry to the file CHANGELOG.adoc

### CODE REVIEW

**Comment by abelsromero** [comment]
Progress report:
* Completed first pass of guides module:
- Changes tiles for the extension guides and first section, and added a preample to each one from the introduction in the previous first sub-section. That's because since the new version split the doc in 2 (1.5 to 1.6, 1.6. to 2.0), I needed a way make the content more fluid.

**Comment by abelsromero** [comment]
* Added SDKman installation descripotion
* Moven Windowd and sdkman sections to CLI page
* Added a link from installation to CLI sections
* Moved Loading Ruby libraries into guides

As discussed in gitter, for now, to solve how to include the code examples that run as tests. I created a gradle task `copyExamplesToDocs` to copy required files into _examples_ dir of the appropiate antora module. Right night very quick-and-dirty but solves the issue. When all is done I will review to improve and remove `processAsciidocFiles` which won't be necessary.

**Comment by robertpanzer** [comment]
Thanks for starting on this @abelsromero, this is a great start!

I see some warnings for missing attributes and some other minor issues.

The extension migration guides could maybe be put into an own section "Updating to new releases" and then subsections "From 1.5 to 1.6", "From 1.6 to 2.0" etc?

For Wildfly and Osgi we could group these into an own section like "Running in frameworks" (there is probably a better title) and then just Wildfly and Osgi as subsections.

**Comment by abelsromero** [comment]
> I see some warnings for missing attributes and some other minor issues.

Still WIP, I updated the title. 

For maven-plugin I tried to get each page 100% done when spliting but I had to redo many things after that. Here I am doing an opposite approach, first split, then review and rework once the overal structure makes sense. That means leaving certain details like attributes or includes.

I know it's kind of a P***, but don't know how to show the progress other than commiting what I get done every day. Ideas are welcome.



> The extension migration guides could maybe be put into an own section "Updating to new releases" and then subsections "From 1.5 to 1.6", "From 1.6 to 2.0" etc?
> 
> For Wildfly and Osgi we could group these into an own section like "Running in frameworks" (there is probably a better title) and then just Wildfly and Osgi as subsections.

Thanks, extra eyes on this things help a lot!

**Comment by mojavelinux** [comment]
Personally, I love receiving the updates. It helps me to appreciate the work you're doing and gives me a reminder to review and give feedback. So I'd say keep doing it the way you are doing it.

I completely emphasize with you about how time consuming editing is. Every time I read one of the new pages in the core docs, I find myself spending several hours reworking and rewriting the content. It's a tough job.

**Comment by mojavelinux** [comment]
You may want to consider making the topic about syntax highlighters a module. And if you want to be consistent with core, we named it "syntax-highlighting" https://asciidoctor-docs.netlify.app/asciidoctor/current/syntax-highlighting/ Using the same structure and terminology will make it easier to remember when making xrefs (e.g., asciidoctor:syntax-highlighting:index.adoc vs asciidoctorj:syntax-highlighting:index.adoc).

**Comment by abelsromero** [comment]
I fixed the imports and missing includes:
- For imports: I had to make some changes in the test  `should_invoke_syntax_highlighter`. The test created a file based on the syntax-highlighter, since this is generated during tests, we cannot get that as static resource to be included by antora.
- For attributes just added them to `antora.yml`.

**Comment by abelsromero** [comment]
> You may want to consider making the topic about syntax highlighters a module.

I think it makes sense to create also an `extensions` module, right now we have 15 pages for extensions in ROOT.

**Comment by abelsromero** [comment]
Only missing points are a couple of sections to integrate and a full review :partying_face:

**Comment by abelsromero** [comment]
* Refactored API options/attributes creation and initialitzation.
There were 2 explanations, one in README and another in the integrators-guide, each one had a different rithm and model. So I refactored it quite a bit. I need to go through a second read however.

* Created new page "examples"
These where used to explain attributes configuration in the README but where not consistent with the new explanations.
So I though it would be nice to extend them a bit and re-purpose them in a new "examples" page where users can see a comple example with options + attributes + convert invocation.

**Comment by abelsromero** [comment]
The files I rename to "_SOMETHING.adoc" is because they contain content that I still need to see if needs to be extracted. Will be deleted in the end.

**Comment by abelsromero** [comment]
Ready. I've made some changes here and there and remove the docs and unncessary docs tasks from the gradle build.
There are improvements to do, but I think is "good enough". More changes in this huge PR is not productive.

To run locally I use.
```
site:
  title: 'Asciidoctorj docs'
content:
  sources:
    - url: .
      branches: [ feature/migrate-docs-to-antora-site ]
      start_path: docs
asciidoc:
  attributes:
    page-pagination: ''
    idprefix: ''
    idseparator: '-'
    experimental: ''
ui:
  bundle:
    url: https://github.com/abelsromero/asciidoctor-maven-plugin-antora-ui/raw/master/bundle/ui-bundle.zip
    snapshot: true
  supplemental_files:
    - path: ui.yml
      contents: |
        static_files: [ .nojekyll ]
    - path: .nojekyll

```

File: docs/antora.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+name: asciidoctorj
+title: Asciidoctorj
```
Comment (mojavelinux) [inline docs/antora.yml:2]: Please set the title of this component version to AsciidoctorJ (uppercase the J).

File: docs/antora.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+version: '2.4'
+asciidoc:
+  attributes:
+    release-version: 2.4.2
+    artifact-version: 2.4.2
+    asciidoctorj-epub3-version: 1.5.0-alpha.19
+    asciidoctorj-pdf-version: 1.5.3
+    jruby-version: 9.2.13.0
+    uri-jruby: http://jruby.org
```
Comment (mojavelinux) [inline docs/antora.yml:11]: Please use url- as a prefix for URLs instead of uri-. The What WG now recommends using URL in favor of URI to avoid confusion.

**Comment by mojavelinux** [comment]
@abelsromero There are a few broken xrefs that are preventing the deploy preview from completing. Could you fix them? See https://app.netlify.com/sites/asciidoctor-docs/deploys/5fc4c5c0d86d675d8785fefb

**Review by robertpanzer — COMMENTED** [review]
This looks great, I don't get any warnings locally (when rendering with Antora 2.3.4.
I just commented a few minor things.

Again, thanks a lot for this big effort!

File: docs/modules/ROOT/pages/cli.adoc [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+To create a PDF with a different name -- say, READTHIS.pdf -- just add the -o switch:
+
+[source]
+[subs="specialcharacters,attributes,callouts"]
+----
+Linux: asciidoctorj-{artifact-version}/bin/asciidoctorj -b pdf -o READTHIS.pdf README.adoc
+Windows: asciidoctorj-{artifact-version}\bin\asciidoctorj.bat -b pdf -o READTHIS.pdf README.adoc
+----
+
+The rest of the document addresses the asciidoctorj API, for doing more complex conversions
```
[the remark this answers is not in the document]
Comment (abelsromero) [reply docs/modules/ROOT/pages/cli.adoc:51]: Thanks! We can just remove this line.
These are the kind of thing I refer as "changes here and there".

File: docs/antora.yml (hunk 1 above)
Comment (abelsromero) [reply docs/antora.yml:2]: I fixed also 2 more occurences of these.

File: docs/antora.yml (hunk 2 above)
Comment (abelsromero) [reply docs/antora.yml:11]: I am not modifying the README for now, at least, not untill we have the site.

**Comment by abelsromero** [comment]
> @abelsromero There are a few broken xrefs that are preventing the deploy preview from completing. Could you fix them? See https://app.netlify.com/sites/asciidoctor-docs/deploys/5fc4c5c0d86d675d8785fefb

I think I am going to create an allias "antora=antora --generator @antora/xref-validator"

File: docs/modules/ROOT/pages/installation.adoc [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+.Declaring the dependency in a Leiningen build file (e.g., project.clj)
+----
+:dependencies [[org.asciidoctor/asciidoctorj "{artifact-version}"]]
+----
+
+[TIP]
+====
+In addition to using AsciidoctorJ directly, you can invoke it as part of your build using the Maven or Gradle plugin.
+// TODO update links when final site is available
+- {uri-maven-guide}[How to Install and Use the Asciidoctor Maven Plugin]
```
[the remark this answers is not in the document]
Comment (abelsromero) [reply docs/modules/ROOT/pages/installation.adoc:54]: Good catch!

File: docs/modules/ROOT/pages/cli.adoc [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+it should display the various run options available (i.e., help).
+
+[source]
+[subs="specialcharacters,attributes,callouts"]
+----
+Linux: asciidoctorj-{artifact-version}/bin/asciidoctorj
+Windows: asciidoctorj-{artifact-version}\bin\asciidoctorj.bat
+----
+
+Next, say you want to convert an ASCIIDOC (.adoc) file -- such as this README -- to a pdf.
```
[the remark this answers is not in the document]
Comment (abelsromero) [reply docs/modules/ROOT/pages/cli.adoc:32]: :facepalm:

File: docs/modules/ROOT/pages/ruby-extensions.adoc [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+.Register a Ruby extension in Java
+----
+RubyExtensionRegistry rubyExtensionRegistry = this.asciidoctor.rubyExtensionRegistry(); // <1>
+rubyExtensionRegistry.loadClass(Class.class.getResourceAsStream("/YellRubyBlock.rb")).block("rubyyell", "YellRubyBlock"); // <2>
+
+String content = asciidoctor.convertFile(new File(
+                "target/test-classes/sample-with-ruby-yell-block.ad"),
+                options().toFile(false).get());
+----
+<1> `rubyExtensionRegistry` method is called to get a rubyExtensionRegistry instance.
```
[the remark this answers is not in the document]
Comment (abelsromero) [reply docs/modules/ROOT/pages/ruby-extensions.adoc:16]: NOTE: not fixing the readme

File: docs/modules/ROOT/pages/read-document-tree.adoc [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+[source,java]
+----
+Document document = asciidoctor.load(DOCUMENT, new HashMap<String, Object>());
+Map<Object, Object> selector = new HashMap<>(); // <1>
+selector.put("context", ":image"); // <2>
+
+List<StructuralNode> findBy = document.findBy(selector);
+assertThat(findBy, hasSize(2)); //<3>
+----
+<1> To make queries you need to use a `map` approach. Currently this is because of the Asciidoctor API but it will change in near future.
```
[the remark this answers is not in the document]
Comment (abelsromero) [reply docs/modules/ROOT/pages/read-document-tree.adoc:56]: I wrote a more neutral _Queries are defined in a `map` estructure_. But remian a problem, we don't have a link to a proper description of accepted keys and values.

File: docs/modules/ROOT/pages/write-converter.adoc [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+Annotate it as a converter for your target format using the annotation `@org.asciidoctor.converter.ConverterFor`.
+- Register the converter at the `ConverterRegistry`.
+- Pass the target format name to the `Asciidoctor` instance when rendering a source file.
+
+A basic converter that converts to an own text format looks like this:
+
+[source,java]
+.org.asciidoctor.converter.TextConverter.java
+----
+include::example$org/asciidoctor/integrationguide/converter/TextConverter.java[tags=include]
```
[the remark this answers is not in the document]
Comment (abelsromero) [reply docs/modules/ROOT/pages/write-converter.adoc:16]: Yes. Double checked the original doc and I saw no other errors.

**Comment by abelsromero** [comment]
All fixes applied! I'd like to call attention to title which are not uniform, should we made them all inifinitive? Also asked on gitter, btw

File: docs/antora.yml (hunk 2 above)
Comment (mojavelinux) [reply docs/antora.yml]: Understood.

**Comment by mojavelinux** [comment]
> I think I am going to create an allias "antora=antora --generator @antora/xref-validator"

That will make it so you can't build the site. The xref validator only validates, then stops. So you may want to name the alias antora-validate or similar.

File: docs/modules/ROOT/pages/read-document-tree.adoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (robertpanzer) [reply docs/modules/ROOT/pages/read-document-tree.adoc]: Nitpick: "map structure" instead of "map estructure"?
I guess "estructure" is spanish? :D

**Comment by robertpanzer** [comment]
For me this looks good.
The question is when we can do the switch?

Maybe we can also create placeholder documents for the previous integrator-guide.adoc which contains the new link for the document?
Users that have bookmarked this document might be confused if they suddenly land on a 404.

**Comment by abelsromero** [comment]
No idea how @mojavelinux did it, but there's a preview https://deploy-preview-23--asciidoctor-docs.netlify.app/asciidoctorj/current/ 🤷

**Comment by abelsromero** [comment]
Now I see, the preview is triggered by https://github.com/asciidoctor/docs.asciidoctor.org/pull/23.
If there's nothing else here, we could merge and update that PR with to point the main branch here.

**Comment by mojavelinux** [comment]
> The question is when we can do the switch?

You can do the switch whenever you're ready. However, the docs won't be live on docs.asciidoctor.org until the site goes live. We're working hard to make that happen asap, but there are still several things that need to be finished. I'm hoping it will be live by early next week. I'm posting progress updates to the chat channel.

**Comment by robertpanzer** [comment]
Can we then simply add the old .adocs from docs/ directory again and remove or replace them once the site is live?
That would imo make the experience seamless.

With that I really mean just to restore the old generated .adocs in docs/, not the source documents under asciidoctorj-documentation.

**Comment by mojavelinux** [comment]
I would just wait to merge until we are going live with the site. It's not going to be that many days, so I don't see any reason why this PR can't just remain open until then.

**Comment by robertpanzer** [comment]
Sure, I just had thought that there was a circular dependency where the site including the AsciidoctorJ docs cannot go online before this PR is merged and I would like to avoid having a time slot where the current documentation is not available.

But I am fine with waiting too.

**Comment by mojavelinux** [comment]
I'll make sure the change over is coordinated so it happens all at once. Until then, you can find a preview of it linked from this PR https://github.com/asciidoctor/docs.asciidoctor.org/pull/23

**Comment by mojavelinux** [comment]
@abelsromero and/or @robertpanzer I'm planning to soft launch docs.asciidoctor.org. That means I will enable docs.asciidoctor.org (so links will resolve), but we won't announce the site until the first week of the new year. Will you be around this week to merge this for the soft launch?

The alternative is that I can point the soft launch to the PR branch and disable the "Edit this Page" link. That will allow us to see the content on the site, but not yet link it back to this repository.

Let me know how you would like to proceed.

**Comment by abelsromero** [comment]
I'll be arround if we need some fixing but @robertpanzer is the one with merge permissions.

**Comment by robertpanzer** [comment]
I'm available too

**Comment by mojavelinux** [comment]
Oh, excellent! I'm happy to say the site is now soft launched. It's currently pulling content from this branch with the edit URL link disabled. If you merge, I'll update the playbook to pull the content from the master branch and re-enable the edit URL link.

**Comment by robertpanzer** [comment]
@mojavelinux I merged the PR, can you please update the playbook?

**Comment by mojavelinux** [comment]
Yep, I'm on it!

**Comment by mojavelinux** [comment]
Done! Welcome to the new docs site. :tada: We'll use the next 1-2 weeks as an opportunity to make any critical content updates before we announce the site after the new year. And, of course, you can make changes anytime after that too.

**Pull request merged**